### PR TITLE
ARSN-300 implement mongoDB logic for lifecycle listings

### PR DESCRIPTION
**Pull request by nicolas2bert**
Introduce a new metadata method: `listLifecycleObject()`

Implement 3 new "delimiters":

- **DelimiterCurrent** used for listing current versions
The Metadata call returns the masters (M prefix) younger than a defined date: `beforeDate`.
No extra filtering action is needed on the Metadata call response.

- **DelimiterNonCurrent** used for listing non-current version: 
The Metadata call returns the versions (V prefix).
The MD response is then processed to only return the non-current versions that became non-current before a defined date: `beforeDate`.

- **DelimiterOrphan** used for listing orphan delete marker:
The Metadata call returns the versions (V prefix).
The MD response is then processed to only return the delete markers with zero noncurrent versions before a defined date: `beforeDate`.

### CODE REVIEW

File: lib/algos/list/delimiterCurrent.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            params.gt = `${DbPrefixes.Master}${this.keyMarker}`;
+        }
+
+        if (this.beforeDate) {
+            params.lastModified = {
+                lt: this.beforeDate,
+            };
+        }
+
+        this.start = Date.now();
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterCurrent.ts:76]: This does not look to be used in this class

File: lib/algos/list/delimiterCurrent.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            params.gte = `${DbPrefixes.Master}${this.prefix}`;
+            params.lt = `${DbPrefixes.Master}${inc(this.prefix)}`;
+        }
+
+        if (this.keyMarker && params.gte && `${DbPrefixes.Master}${this.keyMarker}` >= params.gte) {
+            delete params.gte;
+            params.gt = `${DbPrefixes.Master}${this.keyMarker}`;
+        }
+
+        if (this.beforeDate) {
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterCurrent.ts:70]: Is there anything different in `DelimiterCurrent` than in `DelimiterMaster`, except the support for `beforeDate`? In which case I strongly recommend to rely on `DelimiterMaster` and only add the support for this new optional parameter. Listing code is tricky and touchy, the less new code we have to add, the better it is :)

**Review by jonathan-gramain — COMMENTED** [review]
This PR should be split, it's too large as is and hard to review. For example, only introduce one listing class and its set of tests at a time.

**Review by jonathan-gramain — COMMENTED** [review]
Also, it's nice to have functional tests, I'd suggest to add unit tests as well, in the vein of what exists for Delimiter, DelimiterMaster and DelimiterVersions, those are more apt to catch specific corner cases that could be hard to catch via functional tests, as we can precisely check inputs and outputs, including for example what the `filter()` function returns.

**Comment by nicolas2bert** [comment]
I am going to split the PR into 3 PRs.
